### ISSUE TITLE
`segfault` with version `2.4.0`

### ISSUE THREAD

**Issue by miguelsousa** [rougier/freetype-py#172]
[Font Bakery](https://github.com/googlefonts/fontbakery) has `freetype-py` as a dependency which is [used on one of its checks](https://github.com/googlefonts/fontbakery/blob/c03667570769f87a020cf8fb4fb8375416ea9b6e/Lib/fontbakery/profiles/universal.py#L1919-L1940).

We've received a [report of that check failing with a `segfault` for variable TrueType fonts](https://github.com/googlefonts/fontbakery/issues/4143) when `freetype-py` version 2.4.0 is installed. The crash doesn't happen with version 2.3.0.

I've ran the `freetype-py` code (snippet below) in isolation but couldn't get it to segfault.

```python
import freetype

face = freetype.Face("SourceSerif4Variable-Italic.ttf")
face.set_char_size(48 * 64)
face.load_char("✅")
```

Could this be related with #169 ?

**Comment by miguelsousa** [comment]
I tried the patch proposed in #171 and the segfault still happens. Commenting out the `Face.__del__` method also didn't fix it.

**Comment by HinTak** [comment]
You do know 2.4.0 is about a decade old, right?  I was adapting 2.6.x for the beginning of fontval, which was 2016, 7 years ago.

**Comment by miguelsousa** [comment]
> You do know 2.4.0 is about a decade old

I'm talking about [**freetype-py** v2.4.0](https://pypi.org/project/freetype-py/2.4.0/) which was release in May 4, 2023, not Python 2.4.0.

**Comment by HinTak** [comment]
I think somebody else also updated the bundled freetype version recently, if you are using wheel or whatever it is called, rather than the system-wide freetype on most Linux machines.

**Comment by miguelsousa** [comment]
I don't have `freetype` installed, and as I mentioned above, rolling back `freetype-py` to v2.3.0 no longer causes the segfault.
Is there a way to tell which version of `freetype` is bundled with each `freetype-py`?

**Comment by miguelsousa** [comment]
Following the [release notes of v2.4.0](https://github.com/rougier/freetype-py/releases/tag/v2.4.0), I found [this commit which updated `freetype`](https://github.com/rougier/freetype-py/pull/164/files) from 2.12.0 to 2.13.0.

**Comment by HinTak** [comment]
You can query from within freetype-py - it is returnee by FT_Version(), or something of similar name.

**Comment by HinTak** [comment]
Some ways of installing freetype-py also build a bundled freetype. The version is controlled by
 https://github.com/rougier/freetype-py/blob/master/setup-build-freetype.py so you might want to check that file's history. Regardless, putting something like FT_Version() inside your python code might be a better way of knowing what version of freetype you are using.

**Comment by HinTak** [comment]
It is just `freetype.version()`, according to the code:
https://github.com/rougier/freetype-py/blob/ccd4f5fca0fb1dd69e40b2966db60139ec1c49da/freetype/__init__.py#L85

**Comment by miguelsousa** [comment]
[The crash only happens on macOS](https://github.com/miguelsousa/openbakery/actions/runs/5354030154).

**Comment by HinTak** [comment]
@miguelsousa Have you worked out it is the freetype switch (or the harfbuzz switch) or something else? Thought I'll ask and see if we can put some fix into a possible 2.5.0 .

**Comment by miguelsousa** [comment]
@HinTak I haven't looked into this at all. I wouldn't know where to start looking.

**Comment by HinTak** [comment]
The first thing would be to identify which commit breaks it - although it is probably one of the freetype/harfbuzz switches? Then you try to switch them individually, for example.

**Comment by HinTak** [comment]
The first step would be to identify the change that gives the segfault. That's just relatively generic git-bisect skill, and suitable for people new to freetype/freetype-py to attempt.

**Comment by miguelsousa** [comment]
@HinTak I have higher priority projects to work on and I don't know if or when I'll get to this.

**Comment by rougier** [comment]
Maybe @hammerscript can investigate using the proposed method.

**Comment by miguelsousa** [comment]
I have proposed a patch upstream. See https://gitlab.freedesktop.org/freetype/freetype/-/issues/1137

## Technical Details

### The Problem

1. Rubik-Italic.ttf's `prep` program executes `NPUSHB[255]`
2. This pushes 255 values onto the interpreter stack
3. Font declares `maxStackElements = 153` in `maxp` table
4. FreeType allocates 153 + 32 = 185 stack slots
5. When bytecode tries to use slot 186+, Stack_Overflow error

### The Fix

Increase safety margin from fixed +32 to dynamic:
- Use 50% more than declared (`maxStackElements / 2`)
- With minimum of 128 extra slots
- For Rubik-Italic: 153 + 128 = 281 slots (enough for 255 values)

### Memory Impact

- Per interpreter context: ~512-1024 bytes
- Negligible for modern systems
- No per-glyph overhead (stack is reused)

**Comment by miguelsousa** [comment]
The upstream patch has been merged. See https://gitlab.freedesktop.org/freetype/freetype/-/commit/93e7335adb2e6ab8ed6f6cf2714c9ef0eb3f3f96

**Comment by rougier** [comment]
Does that mean problem fixed here?

**Comment by miguelsousa** [comment]
> Does that mean problem fixed here?

No. The upstream fix still needs to be integrated and a new `freetype-py` version must be released.
